File: projectview/src/main/java/org/jetbrains/bsp/bazel/projectview/model/sections/ProjectViewSingletonSection.java
package org.jetbrains.bsp.bazel.projectview.model.sections;

import java.util.Objects;

public abstract class ProjectViewSingletonSection<T> extends ProjectViewSection {

  protected final T value;

  protected ProjectViewSingletonSection(String sectionName, T value) {
    super(sectionName);
    this.value = value;
  }

  public T getValue() {
    return value;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (!(o instanceof ProjectViewSingletonSection)) return false;
    ProjectViewSingletonSection<?> that = (ProjectViewSingletonSection<?>) o;
    return value.equals(that.value) && sectionName.equals(that.sectionName);
  }

  @Override
  public int hashCode() {
    return Objects.hash(value, sectionName);
  }
}

File: executioncontext/projectview/src/main/java/org/jetbrains/bsp/bazel/projectview/model/sections/ProjectViewSingletonSection.kt
package org.jetbrains.bsp.bazel.projectview.model.sections

import java.nio.file.Path

sealed class ProjectViewSingletonSection<T>(sectionName: String) : ProjectViewSection(sectionName) {
    abstract val value: T
}

data class ProjectViewDeriveTargetsFromDirectoriesSection(override val value: Boolean) :
        ProjectViewSingletonSection<Boolean>(SECTION_NAME) {
    companion object {
        const val SECTION_NAME = "derive_targets_from_directories"
    }
}

data class ProjectViewBazelPathSection(override val value: Path) :
    ProjectViewSingletonSection<Path>(SECTION_NAME) {
    companion object {
        const val SECTION_NAME = "bazel_path"
    }
}

data class ProjectViewBuildManualTargetsSection(override val value: Boolean) :
    ProjectViewSingletonSection<Boolean>(SECTION_NAME) {
    companion object {
        const val SECTION_NAME = "build_manual_targets"
    }
}

data class ProjectViewImportDepthSection(override val value: Int) :
    ProjectViewSingletonSection<Int>(SECTION_NAME) {
    companion object {
        const val SECTION_NAME = "import_depth"
    }
}
